cmd/openshift-tuned.go

```diff
@@ -290,14 +366,65 @@ func nodeChangeHandler(event watch.Event, tuned *tunedState) {
 	// Node labels didn't change, event didn't modify labels
 }
 
-// Keep things simple and monitor pod-wide label changes only.  However, note that a
-// pod-wide label change doesn't necessarily mean tuned daemon has to be reloaded as
-// labels of other pods running on the same node may already have the same labels.
+// podLabelsUnique goes through pod labels of all the pods on a node (podLabelsNodeWide) and
+// returns a subset of podLabels unique to podNsName; i.e. the retuned labels (key & value)
+// will not exist on any other pod that live on the same node as podNsName.
+func podLabelsUnique(podLabelsNodeWide map[string]map[string]string, podNsName string, podLabels map[string]string) map[string]string {
+	unique := map[string]string{}
+
+	if podLabelsNodeWide == nil {
+		return podLabels
+	}
+
+LoopNeedle:
+	for kNeedle, vNeedle := range podLabels {
+		for kHaystack, vHaystack := range podLabelsNodeWide {
+			if kHaystack == podNsName {
+				// Skip the podNsName labels which are part of podLabelsNodeWide
+				continue
+			}
+			if v, ok := vHaystack[kNeedle]; ok && v == vNeedle {
+				// We've found a matching key/value pair label in vHaystack, kNeedle/vNeedle is not unique
+				continue LoopNeedle
+			}
+		}
+
+		// We've found label kNeedle with value vNeedle unique to pod podNsName
+		unique[kNeedle] = vNeedle
+	}
+
+	return unique
+}
+
+// podLabelsNodeWideChange returns true, if the change in current pod labels (podLabels) affects pod labels node-wide.
+// In other words, new or removed labels (key & value) on podNsName does *not* exist on any other pod that lives on the
+// same node as podNsName.
+func podLabelsNodeWideChange(podLabelsNodeWide map[string]map[string]string, podNsName string, podLabels map[string]string) bool {
+	if podLabelsNodeWide == nil {
+		return podLabels != nil && len(podLabels) > 0
+	}
+
+	// Fetch old labels for pod podNsName, not found on any other pod that lives on the same node
+	oldPodLabelsUnique := podLabelsUnique(podLabelsNodeWide, podNsName, podLabelsNodeWide[podNsName])
+	// Fetch current labels for pod podNsName, not found on any other pod that lives on the same node
+	curPodLabelsUnique := podLabelsUnique(podLabelsNodeWide, podNsName, podLabels)
+	// If there is a difference between old and current unique pod labels, a unique pod label was added or removed
+	change := !reflect.DeepEqual(oldPodLabelsUnique, curPodLabelsUnique)
+
+	glog.V(1).Infof("Pod (%s) labels changed node wide: %v", podNsName, change)
+	return change
+}
+
+// podChangeHandler handles pod events.  It ensures information about pod
```

Comment:
Can you please keep your line width for comments similar between the different functions?

Reply:
```
// podChangeHandler handles pod events.  It ensures pod label changes that
// affect the reloading of tuned are recorded so that it can be acted upon later.
// Labels of pods that co-exist on the same node are taken into account to
// avoid unnecessary tuned reloads.
```